### PR TITLE
rails 4 routemon updates

### PR DESCRIPTION
**Pull request by jonathan-fantham**

### CODE REVIEW

File: lib/ruote/mon/storage.rb [hunk 1]
```diff
@@ -53,11 +53,14 @@ def initialize(mongo_db, options={})
         # can't do that...
 
       (TYPES - %w[ msgs schedules ]).each do |t|
-        collection(t).ensure_index('_wfid')
-        collection(t).ensure_index([ [ '_id', 1 ], [ '_rev', 1 ] ])
+        collection(t).indexes.create_one('_wfid' => 1)
+        collection(t).indexes.create_many([
+          { key: {'_id' => 1}},
+          { key: {'_rev' => 1}}
```
Comment (cbandy) [inline lib/ruote/mon/storage.rb]: I'll check tomorrow, but I think this is supposed to be one index over two columns.

File: lib/ruote/mon/storage.rb [hunk 2]
```diff
@@ -108,11 +107,12 @@ def put(doc, opts={})
...
 
       r = begin
-        collection(doc).update(
-          { '_id' => doc['_id'], '_rev' => original['_rev'] },
+        c = collection(doc).find({ '_id' => doc['_id'], '_rev' => original['_rev'] })
+        c.update_one(
           to_mongo(opts[:update_rev] ? Ruote.fulldup(doc) : doc),
-          :safe => true, :upsert => original['_rev'].nil?)
-      rescue Mongo::OperationFailure
+          :safe => true, :upsert => original['_rev'].nil?
```
Comment (cbandy) [inline lib/ruote/mon/storage.rb]: I don't expect `:safe` is an option in newer versions. Should this be a write concern (`:w`?) instead?

File: lib/ruote/mon/storage.rb [hunk 3]
```diff
@@ -203,7 +202,6 @@ def add_type(type)
     # Nukes a db type and reputs it(losing all the documents that were in it).
     #
     def purge_type!(type)
-
```
Comment (cbandy) [inline lib/ruote/mon/storage.rb:206]: I'm on the fence but... it may be good to avoid whitespace changes.

File: lib/ruote/mon/storage.rb [hunk 4]
```diff
@@ -273,8 +271,7 @@ def paginate(cursor, opts)
 
       return cursor.count if opts['count']
 
-      cursor.sort(
-        '_id', opts['descending'] ? Mongo::DESCENDING : Mongo::ASCENDING)
+      cursor.sort(:"_id" => opts['descending'] ? Mongo::Index::DESCENDING : Mongo::Index::ASCENDING)
```
Comment (cbandy) [inline lib/ruote/mon/storage.rb]: These constants are a little awkward here. Let's change this to `opts['descending'] ? -1 : 1`.

File: lib/ruote/mon/storage.rb (hunk 1 above)
Comment (cbandy) [inline lib/ruote/mon/storage.rb]: Yes, this needs to result in a single index over `_id` and `_rev`:

```
> db.expressions.getIndexes();
[
        {
                "v" : 1,
                "key" : {
                        "_id" : 1
                },
                "ns" : "development.expressions",
                "name" : "_id_"
        },
        {
                "v" : 1,
                "key" : {
                        "_wfid" : 1
                },
                "ns" : "development.expressions",
                "name" : "_wfid_1"
        },
        {
                "v" : 1,
                "key" : {
                        "_id" : 1,
                        "_rev" : 1
                },
                "ns" : "development.expressions",
                "name" : "_id_1__rev_1"
        }
]
```

Comment (jonathan-fantham) [inline lib/ruote/mon/storage.rb]: okay thanks for checking that. I'll update that.

File: lib/ruote/mon/storage.rb (hunk 2 above)
Comment (jonathan-fantham) [inline lib/ruote/mon/storage.rb]: I was thinking that too, but I hadn't had the chance to look into it just yet.

File: lib/ruote/mon/storage.rb (hunk 4 above)
Comment (jonathan-fantham) [inline lib/ruote/mon/storage.rb]: yeah sounds good. I wondered briefly why they used constants from another library. I guess they wanted tight coupling deliberately.

**Pull request closed without merging**